src/main/kotlin/uk/gov/justice/digital/hmpps/hmppschallengesupportinterventionplanapi/entity/Referral.kt

```diff
@@ -42,65 +51,196 @@ class Referral(
     referencedColumnName = "record_id",
   ) val csipRecord: CsipRecord,
 
-  @Column(nullable = false) val incidentDate: LocalDate,
+  incidentDate: LocalDate,
+  incidentTime: LocalTime? = null,
+  referredBy: String,
 
-  @Column val incidentTime: LocalTime? = null,
+  @Column(nullable = false) val referralDate: LocalDate,
 
-  @Column(nullable = false, length = 240) val referredBy: String,
+  proactiveReferral: Boolean? = null,
+  staffAssaulted: Boolean? = null,
+  assaultedStaffName: String? = null,
 
-  @Column(nullable = false) val referralDate: LocalDate,
+  @Column val releaseDate: LocalDate? = null,
 
-  @Column val proactiveReferral: Boolean? = null,
+  descriptionOfConcern: String?,
+  knownReasons: String?,
+  otherInformation: String? = null,
+  saferCustodyTeamInformed: OptionalYesNoAnswer,
+  referralComplete: Boolean? = null,
 
-  @Column val staffAssaulted: Boolean? = null,
+  referralCompletedBy: String? = null,
+  referralCompletedByDisplayName: String? = null,
+  referralCompletedDate: LocalDate? = null,
 
-  @Column val assaultedStaffName: String? = null,
+  incidentType: ReferenceData,
+  incidentLocation: ReferenceData,
+  refererAreaOfWork: ReferenceData,
+  incidentInvolvement: ReferenceData?,
+) {
+  @OneToOne(
+    mappedBy = "referral",
+    cascade = [CascadeType.PERSIST, CascadeType.MERGE, CascadeType.REMOVE],
+  )
+  private var decisionAndActions: DecisionAndActions? = null
 
-  @Column val releaseDate: LocalDate? = null,
+  fun decisionAndActions() = decisionAndActions
 
-  @Column val descriptionOfConcern: String?,
+  @PostLoad
+  internal fun clearPropertyChanges() {
+    propertyChanges = mutableSetOf()
+  }
 
-  @Column val knownReasons: String?,
+  @Transient
+  private var propertyChanges: MutableSet<PropertyChange> = mutableSetOf()
 
-  @Column val otherInformation: String? = null,
+  internal fun propertyChanges(): Set<PropertyChange> = propertyChanges.toSet()
 
-  @Enumerated(EnumType.STRING) @Column val saferCustodyTeamInformed: OptionalYesNoAnswer,
+  private fun listenForChanges(name: String, old: Any?, new: Any?) {
+    if (old != new) {
+      propertyChanges.add(PropertyChange(name, old, new))
+    }
+  }
 
-  @Column val referralComplete: Boolean? = null,
+  private fun listenForRdChanges(name: String, old: ReferenceData?, new: ReferenceData?) {
+    if (old?.code != new?.code) {
+      propertyChanges.add(PropertyChange(name, old?.code, new?.code))
+    }
+  }
 
-  @Column(length = 32) val referralCompletedBy: String? = null,
+  var incidentDate: LocalDate = incidentDate
+    set(value) {
+      listenForChanges("incidentDate", field, value)
```

Comment:
Would using `Referral::incidentDate.name` be better than the string? Reflection == generally bad I know and a rename should pick up this string.